enterprise/internal/insights/store/insight_store.go

```diff
@@ -96,11 +94,94 @@ func (s *InsightStore) GetMapped(ctx context.Context, args InsightQueryArgs) ([]
 	return results, nil
 }
 
+func (s *InsightStore) InsertDirtyQuery(ctx context.Context, series *types.InsightSeries, query *types.DirtyQuery) error {
+	q := sqlf.Sprintf(insertDirtyQuerySql, series.ID, query.Query, query.Reason, query.ForTime, s.Now())
+	return s.Exec(ctx, q)
+}
+
+// GetDirtyQueries returns up to 100 dirty queries for a given insight series.
+func (s *InsightStore) GetDirtyQueries(ctx context.Context, series *types.InsightSeries) ([]*types.DirtyQuery, error) {
+	// We are going to limit this for now to some fixed value, and in the future if necessary add pagination.
+	limit := 100
+	q := sqlf.Sprintf(getDirtyQueriesSql, series.ID, limit)
+	return scanDirtyQueries(s.Query(ctx, q))
+}
+
+func scanDirtyQueries(rows *sql.Rows, queryErr error) (_ []*types.DirtyQuery, err error) {
+	if queryErr != nil {
+		return nil, queryErr
+	}
+	defer func() { err = basestore.CloseRows(rows, err) }()
+
+	results := make([]*types.DirtyQuery, 0)
+	for rows.Next() {
+		var temp types.DirtyQuery
+		if err := rows.Scan(
+			&temp.ID,
+			&temp.Query,
+			&temp.Reason,
+			&temp.ForTime,
+			&temp.DirtyAt,
+		); err != nil {
+			return nil, err
+		}
+		results = append(results, &temp)
+	}
+	return results, nil
+}
+
+// GetDirtyQueriesAggregated returns aggregated information about dirty queries for a given series.
+func (s *InsightStore) GetDirtyQueriesAggregated(ctx context.Context, seriesID string) ([]*types.DirtyQueryAggregate, error) {
+	q := sqlf.Sprintf(getDirtyQueriesAggregatedSql, seriesID)
+	return scanDirtyQueriesAggregated(s.Query(ctx, q))
+}
+
+func scanDirtyQueriesAggregated(rows *sql.Rows, queryErr error) (_ []*types.DirtyQueryAggregate, err error) {
+	if queryErr != nil {
+		return nil, queryErr
+	}
+	defer func() { err = basestore.CloseRows(rows, err) }()
+
+	results := make([]*types.DirtyQueryAggregate, 0)
+	for rows.Next() {
+		var temp types.DirtyQueryAggregate
+		if err := rows.Scan(
+			&temp.Count,
+			&temp.ForTime,
+			&temp.Reason,
+		); err != nil {
+			return nil, err
+		}
+		results = append(results, &temp)
+	}
+	return results, nil
+}
+
+const insertDirtyQuerySql = `
+-- source: enterprise/internal/insights/store/insight_store.go:InsertDirtyQuery
+INSERT INTO insight_dirty_queries (insight_series_id, query, reason, for_time, dirty_at)
+VALUES (%s, %s, %s, %s, %s);
+`
+
+const getDirtyQueriesSql = `
+-- source: enterprise/internal/insights/store/insight_store.go:GetDirtyQueries
+select id, query, reason, for_time, dirty_at from insight_dirty_queries
+where insight_series_id = %s
+limit %s;`
+
+const getDirtyQueriesAggregatedSql = `
+-- source: enterprise/internal/insights/store/insight_store.go:GetDirtyQueriesAggregated
+select count(*) as count, for_time, reason from insight_dirty_queries
+where insight_dirty_queries.insight_series_id = (select id from insight_series where series_id = %s)
```

Comment:
```suggestion
join insight_series s on s.id = insight_series_id where s.series_id = %s)
```

Not sure if something like this produces a better query plan.

Reply:
This is small enough that I'm not worried either way, if down the line it becomes an issue we can do some optimizations.